### ISSUE TITLE
HCA CLI seeing MemoryError during upload phase of Cromwell workflow

### ISSUE THREAD

**Issue by sampierson** [HumanCellAtlas/upload-service#372]
I'm transferring this discussion from Slack to Git/ZenHub.
Saman said:

Hi! I re-ran one of the workflows that was getting killed in the upload phase with a config change to try and narrow down the cause, and now I'm seeing an upload exception. Can someone help me understand why this file failed to upload? The upload area is `8a767e59-2d25-4312-ada4-9235eea29c95`
```
Traceback (most recent call last):
  File "/usr/local/bin/hca", line 10, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.6/site-packages/hca/cli.py", line 131, in main
    result = parsed_args.entry_point(parsed_args)
  File "/usr/local/lib/python3.6/site-packages/hca/upload/cli/upload_command.py", line 56, in __init__
    dcp_type="data")
  File "/usr/local/lib/python3.6/site-packages/hca/upload/upload_area.py", line 133, in upload_files
    raise UploadException(error)
hca.upload.exceptions.UploadException: 
The following files failed:
/cromwell_root/broad-dsde-mint-staging-cromwell-execution/caas-cromwell-executions/AdapterOptimus/1a3b9068-9f54-46e3-a869-91a7a78c0513/call-analysis/Optimus.Optimus/2aee7e38-18e1-4f84-bc94-da74a0a4cb68/call-MergeSorted/merged.bam: [Exception] 
Please retry or contact an hca administrator at data-help@humancellatlas.org for help.
```

**Comment by sampierson** [comment]
Sam Pierson [2:16 PM]
@Saman Ehsan What version of the dcp-cli are you using? (edited) 

Saman Ehsan [2:18 PM]
We're using version 4.10.0 here

Sam Pierson   [1 day ago]
@Saman Ehsan okay so here are my findings (in thread)

Sam Pierson   [1 day ago]
I found records of that upload area’s creation in our database, Lambda and API Gateway logs. (edited)

Sam Pierson   [1 day ago]
I found a record of the CLI requesting credentials for upload (successfully) in Lambda and API Gateway logs. (edited)

Sam Pierson   [1 day ago]
There are no files in the upload area, nor record of them in the database, not any mention of this upload area of files in it ever again in logs.

Sam Pierson   [1 day ago]
The server side here is a bust.  So I started looking at the client side.

Sam Pierson   [1 day ago]
Unfortunately the code where that exception is raised is pretty uninformative - it’s collecting status from asynchronously running threads.

Sam Pierson   [1 day ago]
And alas is seems that right now we’re not capturing much from those threads other than “the upload failed”.

Sam Pierson   [1 day ago]
I want to dig into that part further.  It appears that there is a real exception happening further down but that the information on it is being lost when it is caught and converted to an UploadException.

Sam Pierson   [1 day ago]
The code does try to propagate the info, but something about that exception is weird.

Sam Pierson   [1 day ago]
I may have a version of the CLI with some debug in it for you tomorrow.

Sam Pierson   [1 day ago]
@Saman Ehsan I have released a new version of the dcp cli - v5.0.1 that has some additional debugging in it at that point.  Please get Cromwell to use that and see what happens when it hits that point.

Saman Ehsan   [1 day ago]
Thanks @Sam Pierson!

Sam Pierson   [1 day ago]
There may be other problems with that CLI release

Saman Ehsan   [1 day ago]
Oh I just read Parth's comments in the integration channel...should I hold off on testing with that version?

Sam Pierson   [1 day ago]
I'm OOO right this second but @Parth Shah is tracking an issue

Parth Shah   [1 day ago]
@Saman Ehsan you should be fine on integration/staging. It appears that the issue with v5.0.1 is only when uploading to production

Saman Ehsan   [1 day ago]
Ok!

Parth Shah   [1 day ago]
I'm rerunning the dcp tests to confirm that

Parth Shah   [1 day ago]
Yup confirmed. Proceed with your testing

Saman Ehsan   [1 day ago]
Cool, so I just tried again and the stack-trace looks the same, but in the stdout I see:
```Starting upload of 1 files to upload area 5637ac0e-ee6b-4d08-9851-4419449067fe

While uploading /cromwell_root/broad-dsde-mint-staging-cromwell-execution/caas-cromwell-executions/AdapterOptimus/c3be3f03-2cf0-4de8-88d8-fbee004a8c68/call-analysis/Optimus.Optimus/e6b67bf5-9177-4d13-a7a8-74c61152df32/call-MergeSorted/merged.bam encountered exception <class 'MemoryError'>():``` 

Saman Ehsan   [1 day ago]
Are there any more details in the upload service? @Sam Pierson

Sam Pierson   [1 day ago]
There won’t be any more details in Upload.  This is all client-side/networking I think.  Sure seems so, now that I see it says `MemoryError`! (edited)

Sam Pierson   [1 day ago]
I wonder what that is!  Let me research that.

Sam Pierson   [1 day ago]
Okay it means out of memory, apparently.

Sam Pierson   [1 day ago]
https://airbrake.io/blog/python-exception-handling/memoryerror (edited)

Sam Pierson   [1 day ago]
https://stackoverflow.com/questions/11283220/memory-error-in-python/37726090#37726090 says _f you get an unexpected MemoryError and you think you should have plenty of RAM available, it might be because you are using a 32-bit python installation._ (edited)

Sam Pierson   [1 day ago]
So the questions become:
1) how much RAM do your containers have?
2) are you running 64-bit Python?

Sam Pierson   [1 day ago]
and 3) do your jobs share containers?  (in AWS Batch they do)

Sam Pierson   [1 day ago]
and 4) how many files are you uploading simultaneously? (I’m guessing not a lot) (edited)

Saman Ehsan   [1 day ago]
Oh that's interesting! Let me check...

Saman Ehsan   [24 hours ago]
1) The failed task has about 3.8GB RAM available and is only using 12.5% of that
2) We're using the python 3.6 base docker image...do you know how we can tell if it is 64-bit?
3) No, a new VM running a new docker container gets started for every task in a workflow
4)  This task is trying to upload 342 files (one at a time)...which seems unusually high even for the Optimus pipeline (edited)

Sam Pierson   [23 hours ago]
I’m leaning towards #4 as the prime suspect.

Sam Pierson   [23 hours ago]
I wonder how the CLI limits parallelism…

Sam Pierson   [23 hours ago]
`DEFAULT_THREAD_COUNT = multiprocessing.cpu_count() * 2`

Sam Pierson   [23 hours ago]
Actually it is clearly memory related so a combination of #1 and #4 perhaps (edited)

Saman Ehsan   [22 hours ago]
So for #1 the task isn't using the maximum amount of available memory so I'm not sure if increasing that will help :thinking_face:

Sam Pierson   [22 hours ago]
There are 2 ways to attack this problem now I think:
a) we should try to measure the amount of memory the task is using,  according to a stackoverflow article I saw we might be able to use `time -v` to do that
b) we could try to reduce the amount of parallelism, add a new option to `hca upload files` to adjust that

Saman Ehsan   [22 hours ago]
We have a script that monitors the cpu, memory and disk usage of each task, and in this case it used 0.47GB/3.8GB RAM (this is the script: https://github.com/broadinstitute/cromwell-tools/blob/master/notebooks/monitoring.sh#L15)
notebooks/monitoring.sh:15
```function get_mem_info() {```
broadinstitute/cromwell-toolsAdded by GitHub

Saman Ehsan   [22 hours ago]
Also I'm not sure if this makes a difference but it fails when it tries to upload the first file

Saman Ehsan   [22 hours ago]
Is there a file size limit for uploads?

Sam Pierson   [21 hours ago]
There shouldn’t be, on the client size.  How big is that `merged.bam`?
Is there a way to get ahold of the set of files it is trying to upload?

Saman Ehsan   [21 hours ago]
Ah okay so the `merged.bam` file is 8.55GB

Saman Ehsan   [21 hours ago]
And yes! We have a list of all the gcloud paths for those files...they're spread out in different directories for the different analysis tasks that produced them

Parth Shah   [20 hours ago]
@Arathi Mani how much memory did client side checksumming take in your tests?

Arathi Mani   [20 hours ago]
I didn't check actually... I didn't think it would be much at all

Sam Pierson   [20 hours ago]
Did we deploy client-side checksumming now?  I was looking (I guess not very hard) for it in the CLI but didn’t see it.

Parth Shah   [20 hours ago]
Yeah it's in there

Sam Pierson   [4 hours ago]
Okay I would to see if we can reproduce what happens in the container.  @Saman Ehsan can you first get us a copy of that BAM file please.  Let’s experiment with that first.  Later we might need all 342 files, but we could just use a large number of random files for now.  I’ll try uploading it from a linux docker container with 4GB RAM.

Saman Ehsan   [2 hours ago]
Ok thanks! Here's the BAM file `gs://optimus_stage_files_inputs/merged.bam`

**Comment by sampierson** [comment]
@samanehsan you seem to be able to reproduce this problem reliably.  Just to confirm: is this issue seems 100% reproducible with a particular dataset?

**Comment by samanehsan** [comment]
Yes, when I've re-run that particular data bundle it keeps failing in that step. These are all of the attempts I've run to debug the issue: https://job-manager.caas-prod.broadinstitute.org/jobs?q=caas-collection-name%3Dlira-staging%26comment%3DTreutlein%26bundle-uuid%3Daed9abdc-da46-4639-86b5-1453e467ae4b

I've also seen this error in 3/1733 SmartSeq2 workflows in the neuron diff dataset in production although I haven't tried re-running them:
- https://job-manager.caas-prod.broadinstitute.org/jobs/7aa8952a-d22d-4de5-aa6c-63c6a337059b?q=caas-collection-name%3Dlira-prod%26project_shortname%3DSingle%2520cell%2520RNAseq%2520characterization%2520of%2520cell%2520types%2520produced%2520over%2520time%2520in%2520an%2520in%2520vitro%2520model%2520of%2520human%2520inhibitory%2520interneuron%2520differentiation.%26status%3DFailed
- https://job-manager.caas-prod.broadinstitute.org/jobs/dc85ab31-7dcd-467d-b6a7-24451444f035?q=caas-collection-name%3Dlira-prod%26project_shortname%3DSingle%2520cell%2520RNAseq%2520characterization%2520of%2520cell%2520types%2520produced%2520over%2520time%2520in%2520an%2520in%2520vitro%2520model%2520of%2520human%2520inhibitory%2520interneuron%2520differentiation.%26status%3DFailed
- https://job-manager.caas-prod.broadinstitute.org/jobs/7be4a172-e76a-469a-a7d0-6b6fc102a616?q=caas-collection-name%3Dlira-prod%26project_shortname%3DSingle%2520cell%2520RNAseq%2520characterization%2520of%2520cell%2520types%2520produced%2520over%2520time%2520in%2520an%2520in%2520vitro%2520model%2520of%2520human%2520inhibitory%2520interneuron%2520differentiation.%26status%3DFailed

**Comment by sampierson** [comment]
I looked on staging and there are plenty of other large files getting uploaded, so id doesn't seem purely based on file size:

```
upload_staging=> select id, name,size,updated_at from file where size > 5000000000 and name != '10241MB_file' order by updated_at desc limit 10;
   id    |         name          |    size     |          updated_at           
---------+-----------------------+-------------+-------------------------------
 1537633 | ISS_h_heart_03.tar.gz |  8175009090 | 2019-04-23 07:44:54.135364+00
 1537632 | ISS_h_heart_02.tar.gz |  9493161706 | 2019-04-23 07:37:07.302107+00
 1537631 | ISS_h_heart_04.tar.gz | 10903467404 | 2019-04-23 06:56:14.122107+00
 1530190 | merged.bam            | 17518046124 | 2019-04-09 22:40:03.03577+00
 1529602 | merged.bam            | 18675171684 | 2019-04-09 20:06:47.295146+00
 1529253 | merged.bam            | 19702113806 | 2019-04-09 19:30:05.236782+00
 1528159 | merged.bam            | 19716108147 | 2019-04-09 05:35:53.229429+00
 1527509 | merged.bam            | 20958648006 | 2019-04-09 04:50:11.314326+00
 1527102 | merged.bam            | 16773349800 | 2019-04-09 03:05:11.043973+00
 1526487 | merged.bam            |  8011222898 | 2019-04-09 01:49:00.863148+00
```

**Comment by sampierson** [comment]
I am going to try recreate the issue in a Linux Docker environment.  I'm assuming that Cromwell is running Docker images, is that correct @samanehsan ?  If so, what image is used in this case?  Is it something Cromwell specific, Secondary Analysis specific, or generic?

**Comment by samanehsan** [comment]
That sounds like a good idea @sampierson! Yes, Cromwell uses the docker image quay.io/humancellatlas/secondary-analysis-pipeline-tools:v0.48.2 to run the task.

**Comment by sampierson** [comment]
* It is reproducible under Docker running quay.io/humancellatlas/secondary-analysis-pipeline-tools:v0.48.2 with 4G, 8G or 12G memory running on a server with 8G of RAM and no swap.
* It is reproducible under Ubuntu 10.04.2 on a server with 8G of RAM.
* It is NOT reproducible under Ubuntu 10.04.2 on a server with 16G of RAM.
* It IS reproducible under Docker running quay.io/humancellatlas/secondary-analysis-pipeline-tools:v0.48.2 with an 8G memory limit running on a server with 16G of RAM, but the error signature is different: hca gets MEMKILLed later, instead of the getting the MemoryError early.

**Comment by sampierson** [comment]
Found the culprit, in `ClientSideChecksumHandler.ChecksumCalculator.compute()`:
```
                with open(self._filename, 'rb') as _file_object:
                    sink.write(_file_object.read())
```
From the Python reference:
_file.read([size]) Read at most size bytes from the file (less if the read hits EOF before obtaining size bytes). If the size argument is negative or omitted, read all data until EOF is reached._

compute is trying to read the entire file into RAM before writing it to the sink.

**Comment by sampierson** [comment]
Fixed in dcp-cli v5.1.0
